### ISSUE TITLE
Setting both password and key auth leads to broken "two-factor" auth that requires the same factor twice

### ISSUE THREAD

**Issue by marvinthepa** [CESNET/netopeer2#1629]
tl;dr;
====

netopeer2-server asks for password (or key) twice if both key and password are configured

detailed description
====

Running on the current master branch.

I don't know if this is expected behavior, but is seems wrong to me:

configuring both a password and a public key for a user:
```xml
<netconf-server xmlns="urn:ietf:params:xml:ns:yang:ietf-netconf-server">
  <listen>
    <endpoints>
      <endpoint>
        <name>default-ssh</name>
        <ssh>
          <ssh-server-parameters>
            <client-authentication>
              <users>
                <user xmlns:xc="urn:ietf:params:xml:ns:netconf:base:1.0">
                  <name>kevin</name>
                  <public-keys>
                    <inline-definition>
                      <public-key xmlns:ct="urn:ietf:params:xml:ns:yang:ietf-crypto-types">
                        <name>key</name>
                        <public-key-format>ct:ssh-public-key-format</public-key-format>
                        <public-key>AAAAB3NzaC1yc2EAAAADAQABAAABgQCiZFUcxGeViIAERT9/WE4VLG9jkDNxOdE8hjFdxdrCzQGD0yXdab/OO8Ps5IYWJXBPh5kjXNbvKRymWv//9O4cW/aMQmxEXsRHgGlzeG/UTqGVWDQnltRrrVmp80/UZe/izd9lljInhxApIVwzeVm88MaW7mvl8edE9fbXsZnSJTAUx+M/xT+jprd/5boZW+XVTA0Fa+ZxIEJlGuRWS5SfTJYuUsjzIMfm9Zc9Q7pSxc1gDPmVD0Ae67Zp3zaPzxfu2i9Lw+OVItjIHTFJ7qPGfW6p9GTTc0NSk5lB7dFwZYBJ0CJtTlXl17rkz0zqsJ2y2iO7W5i4mBhti/d2pLqnaA6gnOGozyKSxdK4PLolG/TG/2XdpLUDlWRdD11GOGctw0hR4wULsKRAB7EVrLWqsiRqjSZh2wgjbBKAyAGKyotCkCWA1N+x7lopMF+It0u5P6fD8wikwJBeYlF46pp6WSrfsJ/RAwLQo8dW/JffQXSpQXU2+GvtvX7v6+JP1kE=</public-key>
                      </public-key>
                    </inline-definition>
                  </public-keys>
                  <password>$6$XNbrXOxw4dq/TYzc$tpTS8mFhO7Y.zwNDBlFtdKTkzBrmd7nXVE6R5G8wAfuUgVSN5M4Wlf4s/INcoJ7/OjSPuSfwGrcrsIO7wPNHL1</password>
                </user>
              </users>
            </client-authentication>
          </ssh-server-parameters>
        </ssh>
      </endpoint>
    </endpoints>
  </listen>
</netconf-server>
```

then using ssh to log in, ssh asks for the password *twice*, but never checks the key:

```bash
$ ssh -v -p 830 $HOST -s netconf
...
debug1: Next authentication method: password
kevin@10.195.4.5's password:
Authenticated using "password" with partial success.
debug1: Authentications that can continue: publickey,password
Permission denied, please try again.
kevin@10.195.4.5's password:
Authenticated to 10.195.4.5 ([10.195.4.5]:830) using "password".
debug1: setting up multiplex master socket
debug1: channel 0: new [/home/msander/.ssh/master-kevin@10.195.4.5:830]
debug1: channel 1: new [client-session]
debug1: Entering interactive session.
debug1: pledge: id
debug1: Sending subsystem: netconf
<hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">...
```

(especially note `Authenticated using "password" with partial success.`, which according to a few search engine hits, suggests two-factor auth that should ask for both key and password (or both key and keyboard-interactive)).

When configuring the two authentication methods not at once, but one after the other (first password, then public key), the same can be reproduced with keys as well:

```bash
$ ssh -i /tmp/bla -v -p 830 $HOST -s netconf
...
debug1: Next authentication method: publickey
debug1: Offering public key: /tmp/bla RSA SHA256:umrks4xiJwjkbtpnZAHseNkx5HqHsrSkvNJxpg1y0+U explicit
debug1: Server accepts key: /tmp/bla RSA SHA256:umrks4xiJwjkbtpnZAHseNkx5HqHsrSkvNJxpg1y0+U explicit
Authenticated using "publickey" with partial success.
debug1: Authentications that can continue: publickey,password
debug1: Offering public key: /tmp/bla RSA SHA256:umrks4xiJwjkbtpnZAHseNkx5HqHsrSkvNJxpg1y0+U explicit
debug1: Server accepts key: /tmp/bla RSA SHA256:umrks4xiJwjkbtpnZAHseNkx5HqHsrSkvNJxpg1y0+U explicit
Authenticated to 10.195.4.5 ([10.195.4.5]:830) using "publickey".
debug1: setting up multiplex master socket
debug1: channel 0: new [/home/msander/.ssh/master-kevin@10.195.4.5:830]
debug1: channel 1: new [client-session]
debug1: Entering interactive session.
debug1: pledge: id
debug1: Sending subsystem: netconf
<hello xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">...
```

As you can see, the /tmp/bla rsa key is sent a second time after `partial success` , as a "two factor auth" that uses the same factor twice (which is not really two factor after all). This will normally not noticed by the user if they use a key without a password (or an ssh-agent that does not ask for confirmation before a key is used).

This then also works when the key is not specified, but will again ask for the password twice.

expected behavior
====

I would expect one of the following behaviors instead:

* the server accepts each form of authentication, depending on what the ssh-client offers - but does not ask for them twice
* the server requests _both_ key _and_ password
* the config with both password and key is illegal

**Comment by marvinthepa** [comment]
I think the problem is here:

https://github.com/CESNET/libnetconf2/blob/devel/src/session_server_ssh.c#L1552-L1555

```c
        if (state->auth_success_count < state->auth_method_count) {
            /* success, but he needs to do another method */
            VRB(session, "User \"%s\" partially authenticated, but still needs to authenticate via the rest of his configured methods.", username);
            ssh_message_auth_reply_success(msg, 1);
        } else {
```

The ssh client is asked to authenticate again, but there is no code to ensure that it does not use the same method again.

I don't know the code, but I guess it might be solved by adding something like

```c
        ssh_set_auth_methods(session->ti.libssh.session, libssh_auth_methods & ~method);
```

**Comment by michalvasko** [comment]
Seems you will solve your own issue, which you are free to. Or you can wait until tomorrow/Thursday when the developer responsible for this code will come back from a vacation and will take a look at this.

**Comment by marvinthepa** [comment]
Thanks @michalvasko, if you think the expert is interested in fixing this, they will probably do a much better job than me.

**Comment by Roytak** [comment]
Hi, I took a look at this and it indeed was bugged, thank you for reporting this. A fix based on your suggestions that actually enforces a two (or more) factor auth when configured has been submitted. Feel free to try it out.
